src/pipewire/pipewire.cc

```diff
@@ -0,0 +1,575 @@
+void PipeWireOutput::drain()
+{
+    pw_thread_loop_lock(m_loop);
+    if (m_buffer_at > 0)
+        pw_thread_loop_timed_wait(m_loop, 2);
+
+    pw_stream_flush(m_stream, true);
+    pw_thread_loop_timed_wait(m_loop, 2);
+    pw_thread_loop_unlock(m_loop);
+}
+
+void PipeWireOutput::flush()
+{
+    pw_thread_loop_lock(m_loop);
+    m_buffer_at = 0;
+    pw_thread_loop_unlock(m_loop);
+    pw_stream_flush(m_stream, false);
+}
+
+void PipeWireOutput::period_wait()
+{
+    // TODO: Implement
+}
+
+int PipeWireOutput::write_audio(const void * data, int length)
+{
+    pw_thread_loop_lock(m_loop);
+
+    if (m_buffer_at == m_buffer_size)
+    {
+        if (pw_thread_loop_timed_wait(m_loop, 1) != 0)
+        {
+            pw_thread_loop_unlock(m_loop);
+            return 0;
+        }
+    }
+
+    auto size = aud::min<size_t>(m_buffer_size - m_buffer_at, length);
+    memcpy(m_buffer + m_buffer_at, data, size);
+    m_buffer_at += size;
+
+    pw_thread_loop_unlock(m_loop);
+    return size;
+}
+
+void PipeWireOutput::close_audio()
+{
+    if (m_stream)
+    {
+        pw_thread_loop_lock(m_loop);
+        m_ignore_state_change = true;
+        pw_stream_disconnect(m_stream);
+        pw_stream_destroy(m_stream);
+        m_ignore_state_change = false;
+        m_stream = nullptr;
+        pw_thread_loop_unlock(m_loop);
+    }
+
+    if (m_loop)
+        pw_thread_loop_stop(m_loop);
+
+    if (m_registry)
+    {
+        pw_proxy_destroy(reinterpret_cast<pw_proxy *>(m_registry));
+        m_registry = nullptr;
+    }
+
+    if (m_core)
+    {
+        pw_core_disconnect(m_core);
+        m_core = nullptr;
+    }
+
+    if (m_context)
+    {
+        pw_context_destroy(m_context);
+        m_context = nullptr;
+    }
+
+    if (m_loop)
+    {
+        pw_thread_loop_destroy(m_loop);
+        m_loop = nullptr;
+    }
+
+    if (m_buffer)
+    {
+        delete[] m_buffer;
+        m_buffer = nullptr;
+    }
+}
+
+bool PipeWireOutput::open_audio(int format, int rate, int channels, String & error)
+{
+    static const struct pw_core_events core_events = {
+        .version = PW_VERSION_CORE_EVENTS,
+        .done = PipeWireOutput::on_core_event_done
+    };
+
+    static const struct pw_registry_events registry_events = {
+        .version = PW_VERSION_REGISTRY_EVENTS,
+        .global = PipeWireOutput::on_registry_event_global
+    };
+
+    static const struct pw_stream_events stream_events = {
+        .version = PW_VERSION_STREAM_EVENTS,
+        .state_changed = PipeWireOutput::on_state_changed,
+        .process = PipeWireOutput::on_process,
+        .drained = PipeWireOutput::on_drained
+    };
+
+    if (!(m_loop = pw_thread_loop_new("pipewire-main-loop", nullptr)))
+    {
+        AUDERR("PipeWireOutput: unable to create main loop\n");
+        return false;
+    }
+
+    if (!(m_context = pw_context_new(pw_thread_loop_get_loop(m_loop), nullptr, 0)))
+    {
+        AUDERR("PipeWireOutput: unable to create context\n");
+        return false;
```

Comment:
Most of the `return false;` paths in this function appear to leak memory/not clean things up properly (e.g. here we leak `m_loop` which was initialized above).

Reply:
Fixed by calling `close_audio()` in these cases, similar to the PulseAudio plugin.